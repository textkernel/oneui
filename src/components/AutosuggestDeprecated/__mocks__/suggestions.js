export const SUGGESTIONS = [
    { name: 'Sun' },
    { name: 'Moon' },
    { name: 'Mercury' },
    { name: 'Venus' },
    { name: 'Earth' },
    { name: 'Mars' },
    { name: 'Jupiter' },
    { name: 'Saturn' },
    { name: 'Neptune' },
    { name: 'Pluto', disabled: true },
];

export const SUGGESTION_TO_STRING = (item) => (item ? item.name : '');
