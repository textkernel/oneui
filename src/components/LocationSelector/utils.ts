import { Loader, LoaderOptions } from '@googlemaps/js-api-loader';
import { CircularMarker } from '../Map';

const METER_TO_KM = 1000;
const METER_TO_MI = 1609.34;

export type LatLngLiteral = google.maps.LatLngLiteral;
export type LatLng = google.maps.LatLng;
export type GeocoderResult = google.maps.GeocoderResult;
export type LocationSelectorAddressComponent = google.maps.GeocoderAddressComponent;

export type LocationSelectorLocation = CircularMarker & {
    id: string;
    place_id: string;
    addressComponents?: LocationSelectorAddressComponent[];
};
export const GOOGLE_API_LIBRARIES = ['places'] as Array<'places'>;

export const initGoogleMapServices = (options: LoaderOptions) => {
    if (window?.google?.maps) {
        return Promise.resolve();
    }
    return new Loader({
        libraries: GOOGLE_API_LIBRARIES,
        ...options,
    }).load();
};

export const getRadiusInMeters = (radius: number, radiusUnits: string) => {
    if (radiusUnits === 'km') {
        return radius * METER_TO_KM;
    }
    return radius * METER_TO_MI;
};

export const findGeoCenterForPlaceId = (placeId: string): Promise<LatLng> => {
    if (!window?.google?.maps) {
        return Promise.reject(new Error('Google maps is not initialized'));
    }
    const { Geocoder } = window.google.maps;
    const geocoder = new Geocoder();
    return new Promise<LatLng>((resolve, reject) => {
        geocoder.geocode({ placeId }, (results, status) => {
            if (status === 'OK') {
                if (results[0]) {
                    resolve(results[0].geometry.location);
                } else {
                    reject(new Error(`No results found when searching for placeId ${placeId}`));
                }
            } else {
                reject(new Error(`Geocoder failed due to: ${status}`));
            }
        });
    });
};

export const convertCoordinatesIntoAddress = (location: LatLngLiteral): Promise<GeocoderResult> => {
    if (!window?.google?.maps) {
        return Promise.reject(new Error('Google maps is not initialized'));
    }
    const { Geocoder } = window.google.maps;
    const geocoder = new Geocoder();
    return new Promise<GeocoderResult>((resolve, reject) => {
        geocoder.geocode({ location }, (results, status) => {
            if (status === 'OK') {
                if (results[0]) {
                    resolve(results[0]);
                } else {
                    reject(
                        new Error(
                            `No results found when searching for coordinates ${location.lat}, ${location.lng}`
                        )
                    );
                }
            } else {
                reject(new Error(`Geocoder failed due to: ${status}`));
            }
        });
    });
};
