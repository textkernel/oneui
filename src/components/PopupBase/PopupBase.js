import * as React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { createPopper } from '@popperjs/core';
import { POPUP_PLACEMENTS, ESCAPE_KEY } from '../../constants';

export class PopupBase extends React.Component {
    constructor(props) {
        super(props);
        const { anchorRef, popupRef } = props;

        this.state = { isOpen: false };
        this.popper = undefined;
        this.anchorRef = anchorRef || React.createRef();
        this.popupRef = popupRef || React.createRef();

        this.setPopupVisibility = this.setPopupVisibility.bind(this);
        this.handleWindowClick = this.handleWindowClick.bind(this);
        this.handleKeyPress = this.handleKeyPress.bind(this);
    }

    componentDidMount() {
        document.addEventListener('click', this.handleWindowClick);
        document.addEventListener('keydown', this.handleKeyPress, false);
    }

    componentDidUpdate(prevProps, prevState) {
        const { isOpen } = this.state;
        if (isOpen && !prevState.isOpen) {
            this.createPopperInstance();
        }
        if (!isOpen && prevState.isOpen) {
            this.destroyPopperInstance();
        }
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.handleWindowClick);
        document.removeEventListener('keydown', this.handleKeyPress, false);
    }

    handleWindowClick(event) {
        const { isOpen } = this.state;

        if (isOpen) {
            if (!this.wasPopupClicked(event) && !this.wasAnchorClicked(event)) {
                this.close();
            }
        }
    }

    handleKeyPress(event) {
        if (event.key === ESCAPE_KEY) {
            this.close();
        }
    }

    setPopupVisibility(shouldBeOpen) {
        const { isOpen } = this.state;
        if (shouldBeOpen !== isOpen) {
            this.setState({ isOpen: shouldBeOpen });
        }
    }

    getArgs() {
        const { isOpen } = this.state;
        return {
            setPopupVisibility: this.setPopupVisibility,
            isOpen,
        };
    }

    wasPopupClicked(event) {
        if (event.composedPath && this.anchorRef.current) {
            return Array.from(event.composedPath()).some((node) => {
                // Must be Element node
                if (node.nodeType === 1) {
                    return this.popupRef.current.contains(node);
                }
                return false;
            });
        }
        return false;
    }

    wasAnchorClicked(event) {
        if (this.anchorRef.current) {
            return this.anchorRef.current.contains(event.target);
        }
        return false;
    }

    close() {
        const { onClose } = this.props;
        const { isOpen } = this.state;

        if (isOpen) {
            if (onClose) {
                onClose();
            }
            this.setState({ isOpen: false });
        }
    }

    createPopperInstance() {
        if (this.anchorRef.current && this.popupRef.current) {
            const { placement, popperOptions } = this.props;
            this.destroyPopperInstance();
            this.popper = createPopper(this.anchorRef.current, this.popupRef.current, {
                placement,
                ...popperOptions,
            });
        }
    }

    destroyPopperInstance() {
        if (this.popper) {
            this.popper.destroy();
            this.popper = undefined;
        }
    }

    renderAnchor() {
        const { anchorRenderer } = this.props;
        const anchorElem = anchorRenderer(this.getArgs());
        return React.cloneElement(anchorElem, { ref: this.anchorRef });
    }

    renderPopup() {
        const { renderInPortal } = this.props;
        const { isOpen } = this.state;
        if (isOpen) {
            const { popupRenderer } = this.props;
            const popupElem = popupRenderer(this.getArgs());
            if (!popupElem) {
                return null;
            }

            const popupElemWithProps = React.cloneElement(popupElem, {
                ref: this.popupRef,
                'data-popup': 'true',
            });

            return renderInPortal
                ? ReactDOM.createPortal(popupElemWithProps, document.body)
                : popupElemWithProps;
        }

        return null;
    }

    render() {
        return (
            <>
                {this.renderAnchor()}
                {this.renderPopup()}
            </>
        );
    }
}

PopupBase.displayName = 'PopupBase';

PopupBase.propTypes = {
    /**
     * Function, that returns an element that triggers popup.
     * It will be called with a single object as argument that contains:
     *      * setPopupVisibility {function} - can be called with true/false to show/hide the popup
     *      * isOpen {boolean} - the current state of the popup
     * NOTE: The returned element should support refForward, but should not have it set.
     *     If you need to access the ref, pass the ref with anchorRef prop (see below)
     */
    anchorRenderer: PropTypes.func.isRequired,
    /**
     * Function, that returns popup element.
     * It will be called with a single object as argument that contains:
     *      * setPopupVisibility {function} - can be called with true/false to show/hide the popup
     *      * isOpen {boolean} - the current state of the popup
     * NOTE: The returned element should support refForward, but should not have it set.
     *     If you need to access the ref, pass the ref with popupRef prop (see below)
     */
    popupRenderer: PropTypes.func.isRequired,
    /** ref object for anchor */
    anchorRef: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    /** ref object for popup */
    popupRef: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    /** placement of the popup dialog relative to anchor */
    placement: PropTypes.oneOf(POPUP_PLACEMENTS),
    /** additional Popper options. See https://popper.js.org/docs/v2/constructors/#options */
    popperOptions: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    /** To render the popup in a portal. Useful if the anchor element has overflow hidden and similar cases */
    renderInPortal: PropTypes.bool,
    /** a function to be called when the popup closes */
    onClose: PropTypes.func,
};

PopupBase.defaultProps = {
    anchorRef: null,
    popupRef: null,
    placement: 'bottom-start',
    renderInPortal: false,
    onClose: null,
    popperOptions: {},
};
