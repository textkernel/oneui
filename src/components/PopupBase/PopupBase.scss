.PopupBase {

    &__popup {
        composes: OneUI-label-text from global
    }
}
