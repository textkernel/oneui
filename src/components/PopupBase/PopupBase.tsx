import * as React from 'react';
import { ReactElement, useEffect, useLayoutEffect } from 'react';
import ReactDOM from 'react-dom';
import { createPopper, Instance } from '@popperjs/core';
import { ESCAPE_KEY, PopupPlacement } from '../../constants';

export type PopupBaseRendererArgs = {
    setPopupVisibility: (isOpen: boolean) => void;
    isOpen: boolean;
};

export type PopupBaseRenderer = ({
    setPopupVisibility,
    isOpen,
}: PopupBaseRendererArgs) => ReactElement | null;

export interface Props {
    /**
     * Function, that returns an element that triggers popup.
     * It will be called with a single object as argument that contains:
     *      * setPopupVisibility {function} - can be called with true/false to show/hide the popup
     *      * isOpen {boolean} - the current state of the popup
     * NOTE: The returned element should support refForward, but should not have it set.
     *     If you need to access the ref, pass the ref with anchorRef prop (see below)
     */
    anchorRenderer: PopupBaseRenderer;
    /**
     * Function, that returns popup element.
     * It will be called with a single object as argument that contains:
     *      * setPopupVisibility {function} - can be called with true/false to show/hide the popup
     *      * isOpen {boolean} - the current state of the popup
     * NOTE: The returned element should support refForward, but should not have it set.
     *     If you need to access the ref, pass the ref with popupRef prop (see below)
     */
    popupRenderer: PopupBaseRenderer;
    /** ref object for anchor */
    anchorRef?: React.RefObject<HTMLElement>;
    /** ref object for popup */
    popupRef?: React.RefObject<HTMLElement>;
    /** placement of the popup dialog relative to anchor */
    placement?: PopupPlacement;
    /** additional Popper options. See https://popper.js.org/docs/v2/constructors/#options */
    popperOptions?: object;
    /** To render the popup in a portal. Useful if the anchor element has overflow hidden and similar cases */
    renderInPortal?: boolean;
    /** a function to be called when the popup closes */
    onClose?: () => void;
}

export const PopupBase: React.FC<Props> = ({
    anchorRenderer,
    popupRenderer,
    anchorRef,
    popupRef,
    placement = 'bottom-start',
    renderInPortal = false,
    onClose,
    popperOptions = {},
}) => {
    const [isOpen, setIsOpen] = React.useState<boolean>(false);
    const anchorRefInit = anchorRef ?? React.createRef<HTMLElement>();
    const popupRefInit = popupRef ?? React.createRef<HTMLElement>();
    let popper: Instance | undefined;

    const close = () => {
        if (isOpen) {
            if (onClose) {
                onClose();
            }
            setIsOpen(false);
        }
    };

    const wasPopupClicked = (event) => {
        if (event.composedPath && anchorRefInit.current && popupRefInit.current) {
            return Array.from(event.composedPath()).some((node) => {
                // Must be Element node
                if ((node as Element).nodeType === 1) {
                    return popupRefInit.current && popupRefInit.current.contains(node as Element);
                }
                return false;
            });
        }
        return false;
    };

    const wasAnchorClicked = (event) => {
        if (anchorRefInit.current) {
            return anchorRefInit.current.contains(event.target);
        }
        return false;
    };

    const handleWindowClick = (event) => {
        if (isOpen) {
            if (!wasPopupClicked(event) && !wasAnchorClicked(event)) {
                close();
            }
        }
    };

    const handleKeyPress = (event) => {
        if (event.key === ESCAPE_KEY) {
            close();
        }
    };

    const destroyPopperInstance = () => {
        if (popper) {
            popper.destroy();
            popper = undefined;
        }
    };

    const createPopperInstance = () => {
        if (anchorRefInit.current && popupRefInit.current) {
            destroyPopperInstance();
            popper = createPopper(anchorRefInit.current, popupRefInit.current, {
                placement,
                ...popperOptions,
            });
        }
    };

    useLayoutEffect(() => {
        if (isOpen) {
            createPopperInstance();
        } else {
            destroyPopperInstance();
        }
    });

    useEffect(() => {
        document.addEventListener('click', handleWindowClick);
        document.addEventListener('keydown', handleKeyPress, false);

        return () => {
            document.removeEventListener('click', handleWindowClick);
            document.removeEventListener('keydown', handleKeyPress, false);
        };
    });

    const setPopupVisibility = (shouldBeOpen) => {
        if (shouldBeOpen !== isOpen) {
            setIsOpen(shouldBeOpen);
        }
    };

    const getArgs = () => ({
        setPopupVisibility,
        isOpen,
    });

    const renderAnchor = () => {
        const anchorElem = anchorRenderer(getArgs());
        // @ts-ignore
        return anchorRefInit && React.cloneElement(anchorElem, { ref: anchorRefInit });
    };

    const renderPopup = () => {
        if (isOpen) {
            const popupElem = popupRenderer(getArgs());
            if (!popupElem) {
                return null;
            }

            const popupElemWithProps = React.cloneElement(popupElem, {
                ref: popupRefInit,
                'data-popup': 'true',
            });

            return renderInPortal
                ? ReactDOM.createPortal(popupElemWithProps, document.body)
                : popupElemWithProps;
        }

        return null;
    };

    return (
        <>
            {renderAnchor()}
            {renderPopup()}
        </>
    );
};

PopupBase.displayName = 'PopupBase';
