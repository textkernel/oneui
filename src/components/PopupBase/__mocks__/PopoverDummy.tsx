import * as React from 'react';
import { Button } from '../../..';

interface Props {
    setPopupVisibility: (isVisible) => void;
}
export const PopoverDummy = React.forwardRef<HTMLDivElement, Props>(
    ({ setPopupVisibility }, ref) => {
        const popupStyle = {
            backgroundColor: '#FFEEDD',
            padding: '10px',
            borderRadius: '5px',
            border: '1px solid #CCC',
            boxShadow: '0 3px 6px 6px rgba(0, 0, 0, 0.05)',
        };
        return (
            <div style={popupStyle} ref={ref} role="group">
                <p>I am the content of pop up.</p>
                <p>Can be a form or anything.</p>
                <Button onClick={() => setPopupVisibility(false)}>Close me!</Button>
            </div>
        );
    }
);

PopoverDummy.displayName = 'Popover';
