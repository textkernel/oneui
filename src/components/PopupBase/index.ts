export { PopupBase, PopupBaseRenderer } from './PopupBase';
