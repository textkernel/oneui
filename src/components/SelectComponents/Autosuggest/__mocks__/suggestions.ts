export const SUGGESTIONS = [
    { name: 'Sun' },
    { name: 'Moon' },
    { name: 'Mercury' },
    { name: 'Venus' },
    { name: 'Earth' },
    { name: 'Mars' },
    { name: 'Jupiter' },
    { name: 'Saturn' },
    { name: 'Neptune' },
    { name: 'Pluto', disabled: true },
];

export type SuggestionsType = (typeof SUGGESTIONS)[number];

export const COMPLEX_SUGGESTIONS = [
    { name: 'Sun', type: 'star' },
    { name: 'Moon', type: 'star' },
    { name: 'Mercury', type: 'planet' },
    { name: 'Venus', type: 'planet' },
    { name: 'Earth', type: 'planet' },
    { name: 'Mars', type: 'planet' },
    { name: 'Jupiter', type: 'planet' },
    { name: 'Saturn', type: 'planet' },
    { name: 'Neptune', type: 'planet' },
];

export const SUGGESTION_TO_STRING = (item) => (item ? item.name : '');
