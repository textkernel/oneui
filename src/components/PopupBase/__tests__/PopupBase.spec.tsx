import React from 'react';
import { render, RenderResult, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { useDocumentEvent } from '../../../utils/testUtils';
import { Button } from '../../..';
import { PopupBase } from '../PopupBase';
import { PopoverDummy } from '../__mocks__/PopoverDummy';
import { ESCAPE_KEY } from '../../../constants';

describe('<PopupBase> that adds basic anchor/popup functionality to rendered components', () => {
    const anchorRendererMock = ({ setPopupVisibility, isOpen }) => (
        <Button onClick={() => setPopupVisibility(!isOpen)}>Toggle popup</Button>
    );
    const popupRendererMock = ({ setPopupVisibility }) => (
        <PopoverDummy setPopupVisibility={setPopupVisibility} />
    );
    let view: RenderResult;
    let user;

    describe('rendering', () => {
        beforeEach(() => {
            view = render(
                <PopupBase anchorRenderer={anchorRendererMock} popupRenderer={popupRendererMock} />
            );
            user = userEvent.setup();
        });

        it('should render with minimal props correctly', () => {
            expect(view.container).toMatchSnapshot();
            expect(screen.queryByRole('group')).not.toBeInTheDocument();
        });

        it('should render popup when requested', async () => {
            // trigger setPopupVisibility(true) through our dummy component
            await user.click(screen.getByRole('button'));

            expect(view.baseElement).toMatchSnapshot();
            expect(screen.getByRole('group')).toBeInTheDocument();
        });

        it('should render popup in portal when requested', async () => {
            view.rerender(
                <PopupBase
                    anchorRenderer={anchorRendererMock}
                    popupRenderer={popupRendererMock}
                    renderInPortal
                />
            );

            // trigger setPopupVisibility(true) through our dummy component
            await user.click(screen.getAllByRole('button', { name: 'Toggle popup' })[0]);

            expect(view.baseElement).toMatchSnapshot();
            expect(screen.getByRole('group')).toBeInTheDocument();
        });

        it('should support no popup content from renderer', async () => {
            view.rerender(
                <PopupBase anchorRenderer={anchorRendererMock} popupRenderer={() => null} />
            );

            // trigger setPopupVisibility(true) through our dummy component
            await user.click(screen.getAllByRole('button')[1]);

            expect(view.container).toMatchSnapshot();
            expect(screen.queryByRole('group')).not.toBeInTheDocument();
        });

        it('should close popup when requested', async () => {
            // trigger setPopupVisibility(true) through our dummy component
            const triggerButton = screen.getByRole('button');
            await user.click(triggerButton);

            expect(screen.queryByRole('group')).toBeInTheDocument();

            await user.click(triggerButton);

            expect(screen.queryByRole('group')).not.toBeInTheDocument();
        });
    });

    describe('click and keydown event handling', () => {
        const onCloseMock = jest.fn();
        const togglePopup = async () => {
            await user.click(screen.getAllByRole('button')[0]);
        };

        beforeEach(() => {
            view = render(
                <PopupBase
                    anchorRenderer={anchorRendererMock}
                    popupRenderer={popupRendererMock}
                    onClose={onCloseMock}
                />
            );
            user = userEvent.setup();
        });

        it('should close open popup if outside is clicked', async () => {
            await togglePopup();

            expect(screen.queryByRole('group')).toBeInTheDocument();

            await togglePopup();

            expect(screen.queryByRole('group')).not.toBeInTheDocument();
        });

        it('should call onClose if outside is clicked', async () => {
            await togglePopup();
            await user.click(document.body);

            expect(onCloseMock).toHaveBeenCalled();
        });

        it('should close open popup on Escape press', async () => {
            await togglePopup();

            expect(screen.getByRole('group')).toBeInTheDocument();

            await user.keyboard(`[${ESCAPE_KEY}]`);

            expect(screen.queryByRole('group')).not.toBeInTheDocument();
        });

        it('should call onClose on Escape press', async () => {
            await togglePopup();

            await user.keyboard(`[${ESCAPE_KEY}]`);

            expect(onCloseMock).toHaveBeenCalled();
        });

        it('should not close open popup if popup is clicked', async () => {
            const clickDocument = useDocumentEvent('click');

            await togglePopup();

            expect(screen.getByRole('group')).toBeInTheDocument();

            await waitFor(() => {
                clickDocument({
                    composedPath: () => [screen.queryByRole('group')],
                });
            });

            expect(screen.queryByRole('group')).toBeInTheDocument();
        });

        it('should not close open popup if button is clicked (ignoring functionality added by the renderer)', async () => {
            const clickDocument = useDocumentEvent('click');
            await togglePopup();

            expect(screen.getByRole('group')).toBeInTheDocument();

            // clicking directly in the element won't trigger global listener, hence we use our magic mock
            // this also ensures that event handlers defined by the renderer prop are not triggered.
            await waitFor(() => {
                clickDocument({
                    composedPath: () => [screen.queryByRole('group')],
                });
            });

            expect(screen.queryByRole('group')).toBeInTheDocument();
        });
    });
});
