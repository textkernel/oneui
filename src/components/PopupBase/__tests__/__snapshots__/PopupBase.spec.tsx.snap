// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PopupBase> that adds basic anchor/popup functionality to rendered components rendering should render popup in portal when requested 1`] = `
<PopupBase
  anchorRenderer={[Function]}
  popupRenderer={[Function]}
  renderInPortal={true}
>
  <Button
    onClick={[Function]}
  >
    <button
      className="Button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Toggle popup
    </button>
  </Button>
  <Portal
    containerInfo={
      <body>
        <div
          style="background-color: rgb(255, 238, 221); padding: 10px; border-radius: 5px; border: 1px solid #ccc; box-shadow: 0 3px 6px 6px rgba(0, 0, 0, 0.05); position: absolute; left: 0px; top: 0px; margin: 0px;"
        >
          <p>
            I am the content of pop up.
          </p>
          <p>
            Can be a form or anything.
          </p>
          <button
            class="Button"
            type="button"
          >
            Close me!
          </button>
        </div>
      </body>
    }
  >
    <Popover
      data-popup="true"
      setPopupVisibility={[Function]}
    >
      <div
        style={
          Object {
            "backgroundColor": "#FFEEDD",
            "border": "1px solid #CCC",
            "borderRadius": "5px",
            "boxShadow": "0 3px 6px 6px rgba(0, 0, 0, 0.05)",
            "padding": "10px",
          }
        }
      >
        <p>
          I am the content of pop up.
        </p>
        <p>
          Can be a form or anything.
        </p>
        <Button
          onClick={[Function]}
        >
          <button
            className="Button"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Close me!
          </button>
        </Button>
      </div>
    </Popover>
  </Portal>
</PopupBase>
`;

exports[`<PopupBase> that adds basic anchor/popup functionality to rendered components rendering should render popup when requested 1`] = `
<PopupBase
  anchorRenderer={[Function]}
  popupRenderer={[Function]}
>
  <Button
    onClick={[Function]}
  >
    <button
      className="Button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Toggle popup
    </button>
  </Button>
  <Popover
    data-popup="true"
    setPopupVisibility={[Function]}
  >
    <div
      style={
        Object {
          "backgroundColor": "#FFEEDD",
          "border": "1px solid #CCC",
          "borderRadius": "5px",
          "boxShadow": "0 3px 6px 6px rgba(0, 0, 0, 0.05)",
          "padding": "10px",
        }
      }
    >
      <p>
        I am the content of pop up.
      </p>
      <p>
        Can be a form or anything.
      </p>
      <Button
        onClick={[Function]}
      >
        <button
          className="Button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Close me!
        </button>
      </Button>
    </div>
  </Popover>
</PopupBase>
`;

exports[`<PopupBase> that adds basic anchor/popup functionality to rendered components rendering should render with minimal props correctly 1`] = `
<PopupBase
  anchorRenderer={[Function]}
  popupRenderer={[Function]}
>
  <Button
    onClick={[Function]}
  >
    <button
      className="Button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Toggle popup
    </button>
  </Button>
</PopupBase>
`;

exports[`<PopupBase> that adds basic anchor/popup functionality to rendered components rendering should support no popup content from renderer 1`] = `
<PopupBase
  anchorRenderer={[Function]}
  popupRenderer={[Function]}
>
  <Button
    onClick={[Function]}
  >
    <button
      className="Button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Toggle popup
    </button>
  </Button>
</PopupBase>
`;
